### PR TITLE
Fix double click to select full words

### PR DESCRIPTION
**Pull request by KleberPF**
Leaving this as a draft because I want some opinions if this is a good approach or not. Heavily inspired by [this](https://github.com/Chatterino/chatterino2/issues/4960#issuecomment-1981668329) comment from @Nerixyz, but adding an id to each element instead of the prev/next approach. Currently works on basic tests, haven't tested more advanced stuff (RTL text, CJK characters, etc)

### CODE REVIEW

**Review by Nerixyz — COMMENTED** [review]
Using IDs for words is a good idea. However, I think the concept shouldn't "leak" beyond `MessageLayout`, i.e. `ChannelView` and `*MessageBuilder` shouldn't know about it.

One bug this implementation has is that it assumes that one `TextElement` has at most one word. This isn't the case. The `EmotePopup` for example, adds an element with multiple words:

https://github.com/Chatterino/chatterino2/blob/b9b1b8cf9cbb33618e445a100f0a9ef6e11a3a9b/src/widgets/dialogs/EmotePopup.cpp#L53-L55

`MessageLayoutContainer` should generate the IDs. When a `TextElement` adds its elements to the container, it generates a new ID for each word by calling something like `container.nextWordID()`. This way, message builders don't need to know about word-IDs. (if you put the next ID after `containsRTL`, it doesn't consume extra space)

Getting the word bounds should hide the concept of word-IDs too. It should have a single method like `layout.getWordBoundsAt(QPoint)` which finds the layout element and finds the bounds of all layout elements with that word-ID (or returns the element's bounds if it has an ID of -1).

**Comment by KleberPF** [comment]
Thanks for the feedback. I think I implemented everything you pointed out, and it seems to be working well. There might be some off by one errors because the messages are kinda scuffed on KDE Plasma

image

File: src/messages/layouts/MessageLayoutContainer.cpp [hunk 1]
```diff
@@ -456,6 +456,40 @@ size_t MessageLayoutContainer::getFirstMessageCharacterIndex() const
...
+    {
+        const auto &element = this->elements_[index];
+        if (element->getWordId() != wordId)
+        {
+            break;
+        }
+
+        wordEnd += element->hasTrailingSpace()
+                       ? element->getSelectionIndexCount() - 1
+                       : element->getSelectionIndexCount();
```
[the remark this answers is not in the document]
Comment (KleberPF) [reply src/messages/layouts/MessageLayoutContainer.cpp]: Now this is done after the loop checking the last element

File: src/widgets/helper/ChannelView.cpp [hunk 2]
```diff
@@ -1817,7 +1817,7 @@ void ChannelView::mouseMoveEvent(QMouseEvent *event)
     if (this->isDoubleClick_ && hoverLayoutElement)
     {
         auto [wordStart, wordEnd] =
-            getWordBounds(layout.get(), hoverLayoutElement, relativePos);
+            layout->getWordBounds(hoverLayoutElement->getWordId());
```
[the remark this answers is not in the document]
Comment (KleberPF) [reply src/widgets/helper/ChannelView.cpp]: Now we choose which logic to use inside `layout->getWordBounds`

**Comment by KleberPF** [comment]
I tested this a bit more and everything seems to be working. Not sure if there is any point in documenting the tests as they would be just screenshots of random selections but let me know if I should. I'll remove the draft as it seems the final version will be something close to this.

File: src/messages/layouts/MessageLayout.cpp [hunk 3]
```diff
@@ -443,12 +443,31 @@ void MessageLayout::deleteCache()
...
+const MessageLayoutElement *MessageLayout::getElementAt(QPoint point) const
 {
     // go through all words and return the first one that contains the point.
     return this->container_.getElementAt(point);
 }
 
+std::pair<int, int> MessageLayout::getWordBounds(QPoint relativePos) const
+{
+    const auto *element = this->getElementAt(relativePos);
+    assert(element != nullptr);
```
Comment (pajlada) [inline src/messages/layouts/MessageLayout.cpp:455]: getElementAt can explicitly return a nullptr if no element is matched at the position
I see the assert might be ok since we call getElementAt right before this, but since we're just now making `getWordBounds`, couldn't it accept a `MessageLayoutElement` already so the `getElementAt` call isn't done twice?

File: src/messages/layouts/MessageLayout.cpp [hunk 4]
```diff
@@ -443,12 +443,31 @@ void MessageLayout::deleteCache()
...
+
+    if (element->getWordId() != -1)
+    {
+        return this->container_.getWordBounds(element);
+    }
```
Comment (pajlada) [inline src/messages/layouts/MessageLayout.cpp:460]: Could you add a comment here for what this significes? It's not clear to me what it means when the element we're looking at has a word id of -1

File: src/messages/layouts/MessageLayoutContainer.cpp [hunk 5]
```diff
@@ -456,6 +457,45 @@ size_t MessageLayoutContainer::getFirstMessageCharacterIndex() const
...
+        const auto &element = this->elements_[index];
+        if (element->getWordId() != hoveredElement->getWordId())
+        {
+            break;
+        }
+
+        wordEnd += element->getSelectionIndexCount();
+    }
+
+    const auto *lastElementInSelection = this->elements_[index - 1].get();
```
Comment (pajlada) [inline src/messages/layouts/MessageLayoutContainer.cpp:490]: Not sure how feasible it is, but could you add a unit test that runs getWordBounds on a MessageLayoutContainer with some edge cases? like 0 elements, 1 element, hoveredElement doesn't exist

Specifically the `index - 1` part to me is scary

**Review by pajlada — COMMENTED** [review]
I pushed some stuff, mostly small things. Could you review the comments I added to make sure I understood the code correctly?

File: src/messages/layouts/MessageLayout.cpp (hunk 3 above)
Comment (KleberPF) [reply src/messages/layouts/MessageLayout.cpp:455]: I did it that way first, but we need `relativePos` later when calling `this->getSelectionIndex(relativePos)`. Don´t know if there is a better way to do this here (unless you want to pass in both parameters)

File: src/messages/layouts/MessageLayout.cpp (hunk 4 above)
Comment (KleberPF) [reply src/messages/layouts/MessageLayout.cpp:460]: -1 basically means that the element can't span across multiple lines (like an emote, for example). I agree the code is not very self explanatory here. Do you want a comment here or should I encapsulate that inside a function like `element->canBeMultiline()` (name is kinda bad but you get the idea)

File: src/messages/layouts/MessageLayoutContainer.cpp (hunk 5 above)
Comment (KleberPF) [reply src/messages/layouts/MessageLayoutContainer.cpp:490]: Don't know if there is a way to create an empty `MessageLayoutContainer`, but I like having a failsafe here. I will add it and see what I can do about the tests

File: src/messages/layouts/MessageLayout.cpp (hunk 4 above)
Comment (pajlada) [reply src/messages/layouts/MessageLayout.cpp:460]: I think a comment here is fine

File: tests/src/MessageLayout.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+#include "widgets/Notebook.hpp"
+#include "widgets/splits/Split.hpp"
+#include "widgets/splits/SplitInput.hpp"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <QDebug>
+#include <QString>
+
+#include <memory>
```
Comment (KleberPF) [inline tests/src/MessageLayout.cpp:26]: Commenting to remember to remove unused headers

File: tests/src/MessageLayout.cpp [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    MessageLayoutTest(const QString &text)
+    {
+        MessageBuilder builder;
+        builder.append(
+            std::make_unique<TextElement>(text, MessageElementFlag::Text));
+        this->layout = std::make_unique<MessageLayout>(builder.release());
+        this->layout->layout(WIDTH, 1, MessageElementFlag::Text, false);
+    }
+
+    MockApplication mockApplication;
```
Comment (KleberPF) [inline tests/src/MessageLayout.cpp:102]: I still don't quite understand how constructing `MockApplication` here makes it so any all to `getIApp` gets the MockApplication. Also this should probably only be done once instead of each time a test is created

File: tests/src/MessageLayout.cpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+}  // namespace
+
+TEST(TextElement, BasicCase)
+{
+    auto test = MessageLayoutTest("aaaaaaaa");
+
+    // Simulate we are clicking on the first word
+    auto point = QPoint(WIDTH / 20, test.layout->getHeight() / 2);
```
Comment (KleberPF) [inline tests/src/MessageLayout.cpp:113]: This is ugly

File: src/messages/layouts/MessageLayoutContainer.cpp (hunk 5 above)
Comment (KleberPF) [reply src/messages/layouts/MessageLayoutContainer.cpp:495]: I battled my way through gtest and somehow managed to make a working test. I left some comments which I think you should take a look at. I'll add more tests and fix some of the stuff pointed out tomorrow

File: src/messages/MessageElement.cpp [hunk 9]
```diff
@@ -454,7 +454,7 @@ TextElement::TextElement(const QString &text, MessageElementFlags flags,
 void TextElement::addToContainer(MessageLayoutContainer &container,
                                  MessageElementFlags flags)
 {
-    auto *app = getApp();
+    auto *app = getIApp();
```
Comment (KleberPF) [inline src/messages/MessageElement.cpp:457]: Is this ok?

File: src/messages/layouts/MessageLayout.cpp (hunk 3 above)
Comment (KleberPF) [reply src/messages/layouts/MessageLayout.cpp]: Done

File: src/messages/layouts/MessageLayout.cpp (hunk 4 above)
Comment (KleberPF) [reply src/messages/layouts/MessageLayout.cpp]: Added

File: src/messages/MessageElement.cpp (hunk 9 above)
Comment (pajlada) [reply src/messages/MessageElement.cpp:457]: Yes :+1:

File: tests/src/MessageLayout.cpp (hunk 7 above)
Comment (pajlada) [reply tests/src/MessageLayout.cpp:102]: The constructor of MockApplication sets a singleton to point to itself (check mocks/EmptyApplication)

File: tests/src/MessageLayout.cpp (hunk 8 above)
Comment (pajlada) [reply tests/src/MessageLayout.cpp:113]: That's fine :+1: This test has already helped a lot

File: tests/src/MessageLayout.cpp (hunk 6 above)
Comment (pajlada) [reply tests/src/MessageLayout.cpp]: I've gone ahead and done this in [`269e77d` (#5243)](https://github.com/Chatterino/chatterino2/pull/5243/commits/269e77d304edfbe444fc9fa73d01c2718330facd)

**Comment by pajlada** [comment]
Testing RTL selections - it's very scuffed. Luckily, it's already scuffed before this PR, which means I don't think it's a blocker.

**Review by pajlada — APPROVED** [review]
